google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractReadContext.java

```diff
@@ -0,0 +1,558 @@
+import com.google.cloud.spanner.Options.ReadOption;
+import com.google.cloud.spanner.SpannerImpl.AbstractResultSet;
+import com.google.cloud.spanner.SpannerImpl.CloseableIterator;
+import com.google.cloud.spanner.SpannerImpl.GrpcResultSet;
+import com.google.cloud.spanner.SpannerImpl.GrpcStreamIterator;
+import com.google.cloud.spanner.SpannerImpl.ResumableStreamIterator;
+import com.google.cloud.spanner.SpannerImpl.SessionImpl;
+import com.google.cloud.spanner.SpannerImpl.SessionTransaction;
+import com.google.cloud.spanner.spi.v1.SpannerRpc;
+import com.google.protobuf.ByteString;
+import com.google.spanner.v1.BeginTransactionRequest;
+import com.google.spanner.v1.ExecuteBatchDmlRequest;
+import com.google.spanner.v1.ExecuteSqlRequest;
+import com.google.spanner.v1.ExecuteSqlRequest.QueryMode;
+import com.google.spanner.v1.PartialResultSet;
+import com.google.spanner.v1.ReadRequest;
+import com.google.spanner.v1.Transaction;
+import com.google.spanner.v1.TransactionOptions;
+import com.google.spanner.v1.TransactionSelector;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Tracing;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Abstract base class for all {@link ReadContext}s + concrete implementations of read-only {@link
+ * ReadContext}s.
+ */
+abstract class AbstractReadContext
+    implements ReadContext, AbstractResultSet.Listener, SessionTransaction {
+  /**
+   * A {@code ReadContext} for standalone reads. This can only be used for a single operation, since
+   * each standalone read may see a different timestamp of Cloud Spanner data.
+   */
+  static class SingleReadContext extends AbstractReadContext {
+    final TimestampBound bound;
+
+    @GuardedBy("lock")
+    private boolean used;
+
+    SingleReadContext(
+        SessionImpl session, TimestampBound bound, SpannerRpc rpc, int defaultPrefetchChunks) {
+      super(session, rpc, defaultPrefetchChunks);
+      this.bound = bound;
+    }
+
+    @GuardedBy("lock")
+    @Override
+    void beforeReadOrQueryLocked() {
+      super.beforeReadOrQueryLocked();
+      checkState(!used, "Cannot use a single-read ReadContext for multiple reads");
+      used = true;
+    }
+
+    @Override
+    @Nullable
+    TransactionSelector getTransactionSelector() {
+      if (bound.getMode() == TimestampBound.Mode.STRONG) {
+        // Default mode: no need to specify a transaction.
+        return null;
+      }
+      return TransactionSelector.newBuilder()
+          .setSingleUse(TransactionOptions.newBuilder().setReadOnly(bound.toProto()))
+          .build();
+    }
+  }
+
+  private static void assertTimestampAvailable(boolean available) {
+    checkState(available, "Method can only be called after read has returned data or finished");
+  }
+
+  static class SingleUseReadOnlyTransaction extends SingleReadContext
```

Comment:
can this be private?

Reply:
No, this class is referenced from the class SessionImpl.